### DISCUSSION TITLE
Is predicted images of registration method have same affine matrices as the affine matrices of fixed images

### DISCUSSION THREAD

**Discussion by JosianeUwineza** [Project-MONAI/MONAI#6313] [Q&A]
To get registration pair(moving and fixed) images and masks(.nii.gz). I took fixed images and masks rotate them by 10 degrees and perform registration using monai LocalNet model. With this data I found that the dice score reached 97%, but when use Monai SaveImage where resample is set to False, the predicted image is at the same position as moving image buy visualizing it .  when I checked affine of fixed image, moving image and predicted image I found that the moving and predicted image has same affine matrices. Is this mean that the monai LocalNet can't register images and mask which is slightly rotated? what I can do ?

Your helps will be much appreciated

Thanks

**Comment by wyli** [comment]
the output meta info was not updated by the network and workflow. I haven't tested it but please try something like `output_image.affine = fixed.affine` before saving `output_image`.

**Comment by JosianeUwineza** [comment]
@ wyli thanks for the suggestion,   But I think if I use output_image.affine = fixed.affine before saving is like cheating,  how am I going to know the registration model registered the image well? I wish there is the way that this affine matrix can be updated during the training so that the model during predict the image with correct affine

**Reply by wyli** [reply]
I think to measure the registration performance you should compute some image-based similarity metrics, instead of relying on the metadata. to update the affine during training -- it should be feasible with the predicted transform parameters as described here https://github.com/Project-MONAI/MONAI/discussions/6128, I'll create a feature request to make the parameters more accessible. 

--
added https://github.com/Project-MONAI/MONAI/issues/6331

**Reply by JosianeUwineza** [reply]
Thanks, @wyli, to compute the similarity metrics to measure the performance of the registration model is a good idea,  but saving the predicted image during registration is essential to ensure visually that the model successfully does what it suppose to do. Also by considering the answer provided in #6128, I can this work even on LocalNet Model?

**Reply by wyli** [reply]
for the localnet, those non-linear transforms won't be precisely described by the affine, we may consider including deformation fields in metadata. all these are not implemented at the moment, once we have clearer user requirements/feature requests, we should start to design the end-to-end metadata-aware image registration workflows. currently, there are only essential components for registration tasks.

**Reply by JosianeUwineza** [reply]
I tried the GlobaNet on the 3D which I rotated by 5 degrees to get the registration pair, the dice score that I am getting is zero, does GlobaNet work on 3D data?

**Reply by wyli** [reply]
yes

**Reply by JosianeUwineza** [reply]
Any recommendation on how to train GlobaNet in a supervised manner?

**Reply by wyli** [reply]
I think if the goal is to train a supervised network to estimate the affine parameters, there's no need to use GlobalNet (and the globalnet doesn't support this kind of training). you can directly create a regressor to predict the transform, like `self.fc_loc` described in the pytorch tutorial https://pytorch.org/tutorials/intermediate/spatial_transformer_tutorial.html.
